base/toml_parser.jl

```diff
@@ -0,0 +1,1099 @@
+using Base: IdSet
+
+if !@isdefined HAVE_DATES
+    error("need to set the variable `HAVE_DATES` to determine if the Dates stdlib is available")
+end
+
+if HAVE_DATES
+    using Dates
+else
+    # In case we do not have the Dates stdlib available
+    # we parse DateTime into these internal structs,
+    # note that these do not do any argument checking
+    struct Date
+        year::Int
+        month::Int
+        day::Int
+    end
+    struct Time
+        hour::Int
+        minute::Int
+        second::Int
+        ms::Int
+    end
+    struct DateTime
+        date::Date
+        time::Time
+    end
+    DateTime(y, m, d, h, mi, s, ms) =
+        DateTime(Date(y,m,d), Time(h, mi, s, ms))
```

Comment:
Would be so much better if we could either use delayed definitions here (Requires.jl?) or make these a parser error (unsupported) rather than re-implementing incompatible structs with the same names.

Reply:
We can't make it a parser error (we need to support arbitrary TOML files) and we want code loading to work without Dates. So this was the best I could come up with.